pilot/pkg/xds/ep_filters.go

```diff
@@ -61,73 +61,101 @@ func (b *EndpointBuilder) EndpointsByNetworkFilter(endpoints []*LocLbEndpointsAn
 			},
 		}
 
-		// Weight (number of endpoints) for the EDS cluster for each remote networks
-		remoteEps := map[network.ID]uint32{}
-		// Calculate remote network endpoints
+		// Create a map to keep track of the gateways used and their aggregate weights.
+		gatewayWeights := make(map[model.NetworkGateway]uint32)
+
+		// Process all of the endpoints.
 		for i, lbEp := range ep.llbEndpoints.LbEndpoints {
-			epNetwork := network.ID(istioMetadata(lbEp, "network"))
-			// This is a local endpoint or remote network endpoint
-			// but can be accessed directly from local network.
-			gatewaysForNetwork := b.push.NetworkManager().GatewaysForNetwork(epNetwork)
-			if identifier.IsSameOrEmpty(string(b.network), string(epNetwork)) || len(gatewaysForNetwork) == 0 {
-				// Copy on write.
-				clonedLbEp := proto.Clone(lbEp).(*endpoint.LbEndpoint)
-				clonedLbEp.LoadBalancingWeight = &wrappers.UInt32Value{
-					Value: uint32(multiples),
-				}
-				lbEndpoints.emplace(clonedLbEp, ep.tunnelMetadata[i])
-			} else {
-				if !b.canViewNetwork(epNetwork) {
-					continue
-				}
-				// cross-network traffic relies on mTLS to be enabled for SNI routing
-				// TODO BTS may allow us to work around this
-				if b.mtlsChecker.isMtlsDisabled(lbEp) {
-					continue
+			// Copy the endpoint in order to expand the load balancing weight.
+			// When multiplying, be careful to avoid overflow - clipping the
+			// result at the maximum value for uint32.
+			weight := b.scaleEndpointLBWeight(lbEp, scaleFactor)
+			lbEp := proto.Clone(lbEp).(*endpoint.LbEndpoint)
+			lbEp.LoadBalancingWeight = &wrappers.UInt32Value{
+				Value: weight,
+			}
+
+			istioEndpoint := ep.istioEndpoints[i]
+			epNetwork := istioEndpoint.Network
+			epCluster := istioEndpoint.Locality.ClusterID
+			gateways := b.selectNetworkGateways(epNetwork, epCluster)
+
+			// Check if the endpoint is directly reachable. It's considered directly reachable if
+			// the endpoint is either on the local network or on a remote network that can be reached
+			// directly from the local network.
+			if b.proxy.InNetwork(epNetwork) || len(gateways) == 0 {
+				// The endpoint is directly reachable - just add it.
+				lbEndpoints.emplace(lbEp, ep.tunnelMetadata[i])
+				continue
+			}
+
+			// If the proxy can't view the network for this endpoint, exclude it entirely.
+			if !b.canViewNetwork(epNetwork) {
+				continue
+			}
+
+			// Cross-network traffic relies on mTLS to be enabled for SNI routing
+			// TODO BTS may allow us to work around this
+			if b.mtlsChecker.isMtlsDisabled(lbEp) {
+				continue
+			}
+
+			// Apply the weight for this endpoint to the network gateways.
+			remainingWeight := weight
+			for remainingWeight > 0 {
+				// Spread the remaining weight across the gateways.
+				weightPerGateway := remainingWeight / uint32(len(gateways))
```

Comment:
How about 3/2? How can it reflect the real proportion
It is confusing how the weight is calculated.

Reply:
All we're doing here is spreading the weight for the endpoint across all gateways. We could simply round robin across each gateway, subtracting 1 from the endpoint's weight each time (in essence, that's what this code is doing). The division here is just a way to quickly determine the portion of the endpoint weight that each gateway should receive. This calculation should never be `0`, since we scale each endpoint weight by the maximum gateways per network. But even if it were, we adjust it to `1` below.